src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideStreamMemoryBasedAsyncOverrides.cs

```diff
@@ -0,0 +1,219 @@
+                {
+                    var diagnostic = CreateDiagnostic(type, readAsyncArrayOverride, symbols.ReadAsyncMemoryMethod);
+                    context.ReportDiagnostic(diagnostic);
+                }
+
+                if (writeAsyncArrayOverride is not null && writeAsyncMemoryOverride is null)
+                {
+                    var diagnostic = CreateDiagnostic(type, writeAsyncArrayOverride, symbols.WriteAsyncMemoryMethod);
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+
+            static Diagnostic CreateDiagnostic(INamedTypeSymbol violatingType, IMethodSymbol arrayBasedOverride, IMethodSymbol memoryBasedMethod)
+            {
+                RoslynDebug.Assert(arrayBasedOverride.OverriddenMethod is not null);
+
+                //  We want to underline the name of the violating type in the class declaration. If the violating type
+                //  is a partial class, we underline all partial declarations.
+
+                return Diagnostic.Create(
+                    Rule,
+                    violatingType.Locations[0],
+                    violatingType.Locations.Skip(1),
+                    violatingType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                    arrayBasedOverride.OverriddenMethod.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                    memoryBasedMethod.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat));
+            }
+        }
+
+        private static bool TryGetRequiredSymbols(Compilation compilation, out RequiredSymbols requiredSymbols)
+        {
+            var int32Type = compilation.GetSpecialType(SpecialType.System_Int32);
+            var byteType = compilation.GetSpecialType(SpecialType.System_Byte);
+            var byteArrayType = compilation.CreateArrayTypeSymbol(byteType);
+            var memoryOfByteType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1)?.Construct(byteType);
+            var readOnlyMemoryOfByteType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1)?.Construct(byteType);
+            var streamType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream);
+            var cancellationTokenType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingCancellationToken);
+
+            if (memoryOfByteType is null || readOnlyMemoryOfByteType is null || streamType is null || cancellationTokenType is null)
+            {
+                requiredSymbols = default;
+                return false;
+            }
+
+            // We know that there are no compiler errors on streamType, so we use 'SingleOrDefault'.
+
+            var readAsyncArrayMethod = GetOverloads(streamType, ReadAsyncName, byteArrayType, int32Type, int32Type, cancellationTokenType).SingleOrDefault();
```

Comment:
Maybe use `FirstOrDefault()` but assert if the length of the array is more than one?

Reply:
I figured since we're looking up methods on framework types (`Stream`) that are in metadata, and therefore could never contain compiler errors. Now that I think about it though, if this analyzer is running against `System.Private.CoreLib`, these framework types could actually contain compiler errors, and we definitely don't want the analyzer to crash. I suppose we could detect whether `streamType` is in source code or metadata, but it's probably not worth the hassle.